src/SuperVault.sol

```diff
@@ -594,11 +682,56 @@ contract SuperVault is BaseStrategy, ISuperVault {
         /// @notice assign remaining weight to the last index
         newWeights[length - 1] = TOTAL_WEIGHT - totalAssignedWeight;
 
-        /// @dev update SV weights
-        SV.weights = newWeights;
-        SV.superformIds = finalSuperformIds;
-        SV.numberOfSuperforms = length;
+        /// @dev update SV data
+        weights = newWeights;
+        superformIds = finalSuperformIds;
+        numberOfSuperforms = length;
 
         emit RebalanceComplete(finalSuperformIds, newWeights);
     }
+
+    /// @notice Changes the whitelist for a Superform ID
+    /// @param superformId The Superform ID to change
+    /// @param isWhitelisted Whether to whitelist or blacklist
+    function _changeSuperformWhitelist(uint256 superformId, bool isWhitelisted) internal {
+        bool currentlyWhitelisted = whitelistedSuperformIds[superformId];
+
+        // Only process if there's an actual change
+        if (currentlyWhitelisted != isWhitelisted) {
+            whitelistedSuperformIds[superformId] = isWhitelisted;
+
+            if (isWhitelisted) {
+                _addToWhitelist(superformId);
+            } else {
+                _removeFromWhitelist(superformId);
+            }
+
+            emit SuperformWhitelisted(superformId, isWhitelisted);
+        }
+    }
+
+    /// @notice Adds a superform ID to the whitelist array
+    /// @param superformId The Superform ID to add
+    function _addToWhitelist(uint256 superformId) internal {
+        whitelistedSuperformIds[superformId] = true;
+        whitelistedSuperformIdArray.push(superformId);
+    }
+
+    /// @notice Removes a superform ID from the whitelist array
+    /// @param superformId The Superform ID to remove
+    function _removeFromWhitelist(uint256 superformId) internal {
+        whitelistedSuperformIds[superformId] = false;
+
+        uint256 length = whitelistedSuperformIdArray.length;
+        // Find and remove the superformId from the array
+        for (uint256 i; i < length; ++i) {
```

Comment:
this can be done without a while loop if `whitelistedSuperformIds[]` containts the location of the `superformId` in `whitelistedSuperformIdArray`.

Could also use https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/utils/structs/EnumerableSet.sol

Reply:
https://github.com/superform-xyz/SuperVaults/pull/32